efi/pcr_profile.go

```diff
@@ -0,0 +1,297 @@
+//
+// There is some variation in behaviour between different platform firmware
+// implementations, and this will generate a policy that is specific for the current
+// host's platform firmware, based on the contents of the TCG event log. This assumes
+// that any events associated with the first boot attempt (other than the
+// EV_EFI_BOOT_SERVICES_APPLICATION containing the image digest) are measured before
+// the EV_SEPARATOR event. There are some inconsistencies in the TCG PC Client PFP
+// spec v1.05r23 here - section 3.3.4.5 mentions a EV_ACTION "Ready to Boot" signal
+// and that the separator must be recorded prior to this event. It also implies
+// EV_EFI_ACTION events associated with each boot attempt occur after the separator.
+// Section 8.2.4 contradicts this - it makes no mention of any EV_ACTION events and says
+// that the EV_EFI_ACTION "Calling EFI Application from Boot Option" event associated
+// with the first boot attempt must be before the separator. In practise, devices we've
+// enabled follow section 8.2.4 when they measure the first EV_EFI_ACTION event (which
+// is optional - firmware should measure a EV_OMIT_BOOT_DEVICE_EVENTS event if they
+// are not measured, although some implementations don't do this either).
+//
+// Note that on supported platforms, the policy generated by this will include any
+// system preparation applications that were executed during the current boot, as long
+// as they are measured as part of the pre-OS environment (before the EV_SEPARATOR
+// event). Note that the inclusion of these makes a policy inherently fragile because
+// it is not possible to pre-generate policy to accomodate updates of these components.
+//
+// If the EV_OMIT_BOOT_DEVICE_EVENTS is not recorded to PCR 4, the platform firmware
+// may perform meaurements of all boot attempts, even if they fail. The generated policy
+// will be invalid if the platform firmware performs boot attempts that subsequently
+// fail before performing a successful attempt, even if the images associated with the
+// successful attempt are included in this policy.
+func WithBootManagerCodeProfile() PCRProfileOption {
+	return pcrProfileSetFlagsOption(bootManagerCodeProfile)
+}
+
+// AddPCRProfile adds a profile defined by the supplied options to the supplied
+// secboot_tpm2.PCRProtectionProfileBranch, using the specified digest algorithm
+// for the PCR digest. The generated profile is defined by the supplied load
+// sequences and options.
+func AddPCRProfile(pcrAlg tpm2.HashAlgorithmId, branch *secboot_tpm2.PCRProtectionProfileBranch, loadSequences *ImageLoadSequences, options ...PCRProfileOption) error {
+	gen := newPcrProfileGenerator(pcrAlg, loadSequences, options...)
+
+	if gen.flags == 0 {
+		return errors.New("must specify a profile to add")
+	}
+
+	log, err := gen.env.ReadEventLog()
+	if err != nil {
+		return xerrors.Errorf("cannot read TCG event log: %w", err)
+	}
+	gen.log = log
```

Comment:
feels like this could be a method on gen

Reply:
I've moved this inside `addPCRProfile` instead.